File: chrome/android/modules/dev_ui/provider/java/src/org/chromium/chrome/modules/dev_ui/DevUiInstallListener.java
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.modules.dev_ui;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.components.module_installer.engine.InstallListener;

/** {@link InstallListener} implementation that uses JNI to propagate install signal to native. */
@JNINamespace("dev_ui")
public class DevUiInstallListener implements InstallListener {
    private long mNativeListener;

    @Override
    public void onComplete(boolean success) {
        if (mNativeListener == 0) return;
        DevUiInstallListenerJni.get().onComplete(mNativeListener, success);
    }

    @CalledByNative
    private DevUiInstallListener(long nativeListener) {
        mNativeListener = nativeListener;
    }

    @CalledByNative
    private void onNativeDestroy() {
        mNativeListener = 0;
    }

    @NativeMethods
    interface Natives {
        void onComplete(long nativeDevUiInstallListener, boolean success);
    }
}

File: chrome/android/modules/dev_ui/provider/dev_ui_install_listener.cc
// Copyright 2019 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/android/modules/dev_ui/provider/dev_ui_install_listener.h"

#include <utility>

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/android/modules/dev_ui/provider/jni_headers/DevUiInstallListener_jni.h"

namespace dev_ui {

// static
DevUiInstallListener* DevUiInstallListener::Create(
    base::OnceCallback<void(bool)> on_complete) {
  return new DevUiInstallListener(std::move(on_complete));
}

void DevUiInstallListener::OnComplete(JNIEnv* env, bool success) {
  std::move(on_complete_).Run(success);
  delete this;
}

DevUiInstallListener::DevUiInstallListener(
    base::OnceCallback<void(bool)> on_complete)
    : j_listener_(Java_DevUiInstallListener_Constructor(
          base::android::AttachCurrentThread(),
          reinterpret_cast<jlong>(this))),
      on_complete_(std::move(on_complete)) {}

DevUiInstallListener::~DevUiInstallListener() {
  Java_DevUiInstallListener_onNativeDestroy(
      base::android::AttachCurrentThread(), j_listener_);
}

}  // namespace dev_ui
